### PR TITLE
Feature/s3 c 2787 put obj retention

### PR DESCRIPTION
**Pull request by dora-korpar**
Implements PUT Object Retention API. Functional tests to follow.

### CODE REVIEW

**Comment by dora-korpar** [comment]
@bert-e reset

File: lib/api/apiUtils/authorization/prepareRequestContexts.js [hunk 1]
```diff
@@ -1,12 +1,18 @@
...
 const { config } = require('../../../Config');
 
-const RequestContext = policies.RequestContext;
-const requestUtils = policies.requestUtils;
+const { RequestContext, requestUtils } = policies;
 let apiMethodAfterVersionCheck;
-const apiMethodWithVersion = { objectGetACL: true, objectPutACL: true,
-    objectGet: true, objectDelete: true, objectPutTagging: true,
-    objectGetTagging: true, objectDeleteTagging: true };
+const apiMethodWithVersion = {
```
Comment (miniscruff) [inline lib/api/apiUtils/authorization/prepareRequestContexts.js:6]: Probably out of the scope of your work, but using a Set here would probably be more natural than an object where all values are ignored.

Comment (dora-korpar) [reply lib/api/apiUtils/authorization/prepareRequestContexts.js:6]: Yeah totally agreed. That would require a bit of refactoring tho, so not gonna include it in this pr, but I've got a running list of cloudserver refactors that I'll add it to :')

**Review by tmacro — APPROVED** [review]

**Comment by dora-korpar** [comment]
@bert-e reset

File: lib/api/objectPutRetention.js [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+const { pushMetric } = require('../utapi/utilities');
+const getReplicationInfo = require('./apiUtils/object/getReplicationInfo');
+const collectCorsHeaders = require('../utilities/collectCorsHeaders');
+const metadata = require('../metadata/wrapper');
+
+const { parseRetentionXml } = s3middleware.retention;
+const REPLICATION_ACTION = 'PUT_RETENTION';
+
+/**
+ * Object Put Retention - Adds retention information to object
```
Comment (miniscruff) [inline lib/api/objectPutRetention.js:17]: Including the name of the function in the function description is not necessary

File: lib/api/objectPutRetention.js [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+
+    const decodedVidResult = decodeVersionId(request.query);
+    if (decodedVidResult instanceof Error) {
+        log.trace('invalid versionId query', {
+            versionId: request.query.versionId,
+            error: decodedVidResult,
+        });
+        return process.nextTick(() => callback(decodedVidResult));
+    }
+    const reqVersionId = decodedVidResult;
```
Comment (miniscruff) [inline lib/api/objectPutRetention.js:37]: Not sure creating a second variable here is helpful, could of just kept using the original, maybe a new name but two copies seems a bit much

**Review by miniscruff — APPROVED** [review]

**Review by ilkescality — APPROVED** [review]

**Comment by dora-korpar** [comment]
@bert-e reset

**Comment by dora-korpar** [comment]
@bert-e approve

**Pull request merged**